### ISSUE TITLE
Integrate resource viewer code

### ISSUE THREAD

**Issue by GlennPruitt** [paranext/paranext-core#60]
As an architect I need to provide an example extension that does something useful.

Develop an extension that appears in the Paranext "Hello World" example.  

Consider using Chris Klapp's Scripture viewer code to display Scripture text in a read only view.

Depends on #44, #59

**Comment by tjcouch-sil** [comment]
The Simple USFM Editor App is a simple editor demo Chris Klapp made. The community-maintained editor based on it is the Xelah editor. The friendsofagape USFM Editor is a Slate-based editor that Autographa made and gave up on. They had a contractor make the editor for them over two years and did not make a lot of progress.

Marc Durdin, Makara Sok, and the Keyman team are running tests on various editor libraries for us:
- Prosemirror: https://github.com/keymanapp/keyman/issues/7902#issuecomment-1350811703
  - Mobile: https://github.com/keymanapp/keyman/issues/7978
- Lexical Text Editor: https://github.com/keymanapp/keyman/issues/7937#issuecomment-1360925872
  - Mobile: https://github.com/keymanapp/keyman/issues/7977
- SlateJS: https://github.com/keymanapp/keyman/issues/7938#issuecomment-1360930262
  - Mobile: https://github.com/keymanapp/keyman/issues/7979
- USX Editor: https://github.com/keymanapp/keyman/issues/8312
  - Mobile: In progress
- Xelah HtmlPerfEditor: https://github.com/keymanapp/keyman/issues/8324
  - Mobile: In progress

The USX Editor seems to be a good baseline for what works in practically just pure HTML contenteditable. There are a few differences between the baseline and tests in these other editors.

**Comment by tjcouch-sil** [comment]
Quick USFM viewer? https://www.npmjs.com/org/xelah

**Comment by katherinejensen00** [comment]
Here is a [document ](https://docs.google.com/document/d/1erjzEfmGhbv3P39g01Q-RsIHSDPMI9A9B8AvSFaHHVk/edit#)that contains some research done into existing scripture utilities. From what I can tell, VerseRef from ScriptureForge is probably what we want to use, but we will want to talk about this as a group.

**Comment by GlennPruitt** [comment]
Katherine,

Thank you, this is very good information.  I appreciate it.

-Glenn

*…………………………………*
[image: Paratext logo]
*Glenn Pruitt*
Paratext Product Manager
***@***.***
***@***.***>

**Comment by katherinejensen00** [comment]
This comment is mostly to capture learning and document helpful details about the journey to get to where we are now before we pivot to doing anything we can to get a resource viewer to work.

We started with the goal of using the Xelah editor in read only mode to display USFM resources. As I began, I learned that Xelah is "a collection of libraries" that is "...focused on ste-perf-html but architected to be useful for other projects needing to edit strings parsable into sections and blocks" according to their GitHub ReadMe. There are editors based on these libraries so the question then became which one to use. Ideally, we would use a pure USFM editor that took in USFM, edited USFM, and round tripped USFM back. 
There is a beta of a [pure USFM editor](https://github.com/xelahjs/xelah/tree/master/packages/pending/%40ster-type-usfm) being developed, but it doesn't have an npm install yet and hasn't had a commit since late July last year. We could potentially package it and install it locally if we ever want to go back and pursue that path.
While talking with developers who work on xelah libraries and related tools about which editor to use, they suggested using the oce-editor. It used proskomma, but should still take in USFM and should be roundtrippable for all intents and purposes except for some white space differences. They were curious if we could come up with some cases where round tripping failed. The oce-editor is actively being worked on and is ready to be imported/installed. For a resource viewer that didn't need roundtripping, this seemed like the quickest way to go to get something working. 
We started with oce-editor-tools/simple and tried to npm install that. We ran into some package dependency conflicts. The oce-editor-tools team gave us a basic boilerplate project that used oce-editor-tools/simple so we could see that it was successfully imported and so we could compare it with our extension installation and see where it was breaking down. Npm install failed in their basic project which had us bewildered until Ira noticed the project had a yarn file as well. Normally both yarn and npm will work, but since npm wasn't working we decided to give yarn install a try. Yarn worked and we were able to use synp on the created yarn file to generate a package-lock file that we could then successfully use npm install on. This worked in the boilerplate app and worked once in paranext-core. However, it was fragile in paranext-core and kept needing to be redone anytime a branch was switched or a change was pulled or merged in. Eventually, that trick stopped working all together. We tried without success to fix the package dependencies.
We went back to the oce-editor-tools team, who has been helpful and quick to respond throughout the process, and after a few attempts to fix it, they suggested using oce-editor-tools/core instead since they aren't really working on simple anymore. That made sense so we switched to that and got it working in the boiler plate app after some syntax changes. When we tried to get it working in parenext-core though we ran into similar package depency problems that we were again struggling to resolve. We were also concerned about the data file that didn't look like typical usfm. Since oce-editor-tools used proskomma to display and edit usfm it is possible it may have transformed the usfm to a certain point and put that in the data file to save a step, but I didn't look into that to confirm. The view takes advantage of proskomma and epitelete-html. Here is a [vocabulary doc](https://docs.google.com/document/d/1bIjrVtk1RTBepmCecuCWegOyYsoKe5DNNkMhsFWS4EQ/edit) with some definitions of terms that was kind of helpful while researching this. The oce-editor-tools/core team is continuing to improve it and are removing some old dependencies pretty soon.
Now we are going to research the quickest way to get a resource viewer extension in paranext-core.
